tests/integration/query.ts

```diff
@@ -50,3 +59,139 @@ it('universal autocomplete sets search intents', async () => {
   expect(statefulCore.state.query.searchIntents).toEqual(['NEAR_ME']);
 });
 
+
+describe('ensure correct results from latest request', () => {
+  jest.useFakeTimers();
+  const mockAutoCompleteFn = jest.fn(
+    async (request: VerticalAutocompleteRequest | UniversalAutocompleteRequest) => {
+      const waitTime = request.input?.length;
+      return new Promise(res => setTimeout(() => res({ results: [ {value: request.input} ]}), waitTime));
+    }
+  );
+
+  const mockedCore: any = {
+    verticalAutocomplete: mockAutoCompleteFn,
+    universalAutocomplete: mockAutoCompleteFn,
+    verticalSearch: jest.fn( async (request: VerticalSearchRequest) => {
+      const waitTime = request.query?.length;
+      return new Promise(res => setTimeout(() => res(
+        { verticalResults: { results: [request.query] } }), waitTime));
+    }),
+    universalSearch: jest.fn( async (request: UniversalSearchRequest) => {
+      const waitTime = request.query?.length;
+      return new Promise(res => setTimeout(() => res(
+        { verticalResults: [{ results: [request.query] }] }), waitTime));
+    })
+  };
+  const stateManager = new ReduxStateManager();
+  const httpManager = new HttpManager();
+  const statefulCore = new StatefulCore(mockedCore, stateManager, httpManager);
+  statefulCore.setVerticalKey('someKey');
+  const queries = ['really long request', 'short', 'long request'];
+  const updateResult = jest.fn();
+
+  beforeEach(() => {
+    jest.clearAllMocks();
+  });
+
+  it('vertical autocomplete get correct results based on up-to-date response', async () => {
+    statefulCore.addListener({
+      valueAccessor: state => state.vertical?.autoComplete?.results,
+      callback: updateResult
+    });
+
+    statefulCore.setQuery(queries[0]);
+    const firstResponsePromise = statefulCore.executeVerticalAutoComplete();
+    statefulCore.setQuery(queries[1]);
+    const secondResponsePromise = statefulCore.executeVerticalAutoComplete();
+    statefulCore.setQuery(queries[2]);
+    const thirdResponsePromise = statefulCore.executeVerticalAutoComplete();
+
+    jest.advanceTimersByTime(queries[1].length);
+    await secondResponsePromise;
+    expect(statefulCore.state.vertical.autoComplete.results).toEqual([{ value: queries[1] }]);
+    jest.advanceTimersByTime(queries[2].length);
```

Comment:
why do we advance from 5ms to 17ms, and then runAllTimers instead of just runAllTimers?